### PR TITLE
Fix PromoteReplica call in ERS

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

As described in https://github.com/vitessio/vitess/issues/15935, running PromoteReplica before calling `SetReplicationSource` on the replicas is an issue as it can cause `PromoteReplica` to indefinitely hang.

This PR fixes 2 issues - 
1. We now run `PromoteReplica` in parallel with `SetReplicationSource`. This also allows us to not run the `PrimaryPosition` RPC separately, since `PromoteReplica` also returns the position of the primary.
2. We should be using a context that has a timeout when running any RPC calls. Previously `PromoteReplica` and `InitPrimary` both were being called without adding a context timeout. This has also been fixed now.

## Related Issue(s)

- Fixes https://github.com/vitessio/vitess/issues/15935

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go:609]: This message doesn't make sense. Maybe it should be just `failed to promote new primary`? And if so, is it worth printing out the alias of the new primary? Which would end up being something like `failed to promote %s to primary` with the tablet alias plugged in.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go:612]: We log a warning, and also return an error? Why do we log it as a warning and not log it at error level?
Please review this error message as well to see whether it makes sense to append the tablet alias to it.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go:720]: This function now seems to be just 1 line. Is it called in multiple places? Can it be eliminated or combined with `reparentReplicas`? We may need a new name for the combined function, because it is effectively `promoteNewPrimaryAndReparentReplicas`.

Comment (deepthi) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go:720]: I guess it's actually `possiblyPromoteNewPrimaryAndDefinitelyReparentReplicas` 😁 I leave it to you to find a more elegant or concise name.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go:527]: Didn't we say that we should eliminate the option of using ERS to initialize a new cluster?

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (deepthi) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go:530]: typically we use RemoteOperationTimeout for RPCs. waitReplicasTimeout is special, it's passed in from the caller. It may be worth creating a separate context to call PromoteReplica with using RemoteOperationTimeout.

File: go/test/endtoend/reparent/newfeaturetest/reparent_test.go
Comment (deepthi) [inline go/test/endtoend/reparent/newfeaturetest/reparent_test.go:175]: Does sidecardb produce `create if not exists` type statements? otherwise, dropping this table on the primary but not on the replicas will mean that a `create table` will fail when replicated to replicas.

**Review by deepthi — COMMENTED** [review]
I have a few comments that I think are important to resolve before this can be merged.
Also, can we create a test case for the context fix? We'd have to write the test without this PR's changes first and make sure it fails, and that it passes with this PR.

File: go/test/endtoend/reparent/newfeaturetest/reparent_test.go
Comment (deepthi) [inline go/test/endtoend/reparent/newfeaturetest/reparent_test.go:162]:
```suggestion
// write that happens when PromoteReplica is called.
```

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go]: Fixed

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go]: Fixed. We log an error via the logger. The logger might choose to ignore it too, so the error is required to be sent back.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go:720]: I removed the function.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go]: I left it in there. Doesn't hurt.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go]: Fixed 👍

File: go/test/endtoend/reparent/newfeaturetest/reparent_test.go
Comment (GuptaManan100) [reply go/test/endtoend/reparent/newfeaturetest/reparent_test.go:175]: Yes, it does 👍

File: go/test/endtoend/reparent/newfeaturetest/reparent_test.go
Comment (GuptaManan100) [reply go/test/endtoend/reparent/newfeaturetest/reparent_test.go]: Fixed

**Review by deepthi — APPROVED** [review]

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (mattlord) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go:525]: Might be worth adding some nil checks here, although I think we catch panics in this path IIRC.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (mattlord) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go:490]: We should add a comment regarding what this parameter means as it's not obvious (to me anyway) from the name.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (mattlord) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go:526]: Log messages are much nicer IMO when using `topoproto.TabletAliasString`

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (mattlord) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go:540]: Probably worth adding the tablet alias here as well?

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (mattlord) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go:614]: You don't generally want to wrap an error AND include the error again in the message you're wrapping the original/parent error with:
```suggestion
		return nil, vterrors.Wrapf(primaryErr, "failed to promote %v to primary", topoproto.TabletAliasString(newPrimaryTablet.Alias))
```

**Review by mattlord — COMMENTED** [review]
Where did this change? It was not apparent to me: 
```
We now run PromoteReplica in parallel with SetReplicationSource
```

I wanted to understand and confirm that part (sorry if I just missed something obvious). Other than that, just minor comments.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go:525]: I considered it, but decided against it, because before in the workflow, we call `ev.ShardInfo = *shardInfo`. When we call this function `ev`, `ev.SharInfo` have to be non-nil. I can still add the checks, but they won't help since they're guaranteed to be non-nil values.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (mattlord) [inline go/vt/vtctl/reparentutil/emergency_reparenter.go:499]: From looking at this function, I can see that we're not using the context that was passed into the function as the parent context here. We should, shouldn't we?

**Comment by mattlord** [comment]
> Thank you for the review @mattlord. I'll work on addressing them 👍.
> 
> As far as the changes for running `PromoteReplica` in parallel, earlier we would call `PromoteReplica` and then the `reparentReplicas` function. Now, we call `PromoteReplica` as part of `handlePrimary` in `reparentReplicas`. `handlePrimary` runs in parallel with `handleReplica` which runs `SetReplicationSource`.

OK, I see. So here: https://github.com/vitessio/vitess/pull/15934/files#diff-aace4cbbbebad37f5072421e16ca464c0f0b4eb98ad287380c2c764d11e7b836R310-R311

We're calling the `reparentReplicas` function — which does the promotion of the new primary and the replica reparents — instead of doing those two things serially. Now that I've looked at that `reparentReplicas` function I can see what you meant (it uses goroutines for each higher level task of promotion and reparent). Thanks!

**Review by mattlord — APPROVED** [review]
I only had minor comments/requests left so approving so that you can merge when you've addressed those as you feel is best.

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go:526]: We call `handlePrimary` with `TabletAliasString`, so the alias we're using for the message is well formatted.

Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go:526]: This is the handlePrimary call - `handlePrimary(topoproto.TabletAliasString(newPrimaryTablet.Alias), newPrimaryTablet)`

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go:499]: Yes good point, I'll fix it.

File: go/test/endtoend/reparent/newfeaturetest/reparent_test.go
Comment (GuptaManan100) [reply go/test/endtoend/reparent/newfeaturetest/reparent_test.go]: Fixed

File: go/vt/vtctl/reparentutil/emergency_reparenter.go
Comment (GuptaManan100) [reply go/vt/vtctl/reparentutil/emergency_reparenter.go]: Fixed

**Comment by GuptaManan100** [comment]
Thank you @mattlord! I've addressed all the review comments, and just waiting for the tests now! 🚀

**Pull request merged**
